nx/source/services/vi.c

```diff
@@ -45,772 +57,315 @@ Result viInitialize(ViServiceType service_type)
     }
 
     if (R_SUCCEEDED(rc)) {
-        struct {
-            u64 magic;
-            u64 cmd_id;
-            u64 inval;
-        } raw;
-
-        raw.magic = SFCI_MAGIC;
-        raw.cmd_id = g_viServiceType; // ViServiceType matches the cmdid needed to open an IApplicationDisplayService session.
-        raw.inval = (g_viServiceType == ViServiceType_Manager || g_viServiceType == ViServiceType_System) ? 1 : 0;
-
-        rc = _viGetSession(&root_srv, &g_viIApplicationDisplayService, &raw, sizeof(raw));
+        const u32 inval = (g_viServiceType == ViServiceType_Manager || g_viServiceType == ViServiceType_System) ? 1 : 0;
+        rc = _viGetSession(&root_srv, &g_viIApplicationDisplayService, g_viServiceType, inval);
         serviceClose(&root_srv);
     }
 
-    if (R_SUCCEEDED(rc))
+    if (R_SUCCEEDED(rc)) {
         rc = _viGetSessionNoParams(&g_viIApplicationDisplayService, &g_viIHOSBinderDriverRelay, 100);
+    }
 
-    if (g_viServiceType >= ViServiceType_System && R_SUCCEEDED(rc))
+    if (g_viServiceType >= ViServiceType_System && R_SUCCEEDED(rc)) {
         rc = _viGetSessionNoParams(&g_viIApplicationDisplayService, &g_viISystemDisplayService, 101);
+    }
 
-    if (g_viServiceType >= ViServiceType_Manager && R_SUCCEEDED(rc))
+    if (g_viServiceType >= ViServiceType_Manager && R_SUCCEEDED(rc)) {
         rc = _viGetSessionNoParams(&g_viIApplicationDisplayService, &g_viIManagerDisplayService, 102);
+    }
 
-    if (g_viServiceType >= ViServiceType_System && R_SUCCEEDED(rc) && hosversionAtLeast(2,0,0))
+    if (g_viServiceType >= ViServiceType_System && R_SUCCEEDED(rc) && hosversionAtLeast(2,0,0)) {
         rc = _viGetSessionNoParams(&g_viIApplicationDisplayService, &g_viIHOSBinderDriverIndirect, 103);
-
-    if (R_FAILED(rc))
-        viExit();
+    }
 
     return rc;
 }
 
-void viExit(void)
-{
-    serviceClose(&g_viIHOSBinderDriverIndirect);
-    serviceClose(&g_viIManagerDisplayService);
-    serviceClose(&g_viISystemDisplayService);
-    serviceClose(&g_viIHOSBinderDriverRelay);
-    serviceClose(&g_viIApplicationDisplayService);
+void _viCleanup(void) {
+    if (serviceIsActive(&g_viIHOSBinderDriverIndirect))   serviceClose(&g_viIHOSBinderDriverIndirect);
```

Comment:
Checking serviceIsActive is unnecessary. serviceClose is a nop when the service isn't active, by design.